powertools-validation/src/main/java/software/amazon/lambda/powertools/validation/ValidationUtils.java

```diff
@@ -229,31 +230,37 @@ public static JsonSchema getJsonSchema(String schema) {
     public static JsonSchema getJsonSchema(String schema, boolean validateSchema) {
         JsonSchema jsonSchema = schemas.get(schema);
 
-        if (jsonSchema == null) {
-            if (schema.startsWith(CLASSPATH)) {
-                String filePath = schema.substring(CLASSPATH.length());
-                InputStream schemaStream = ValidationAspect.class.getResourceAsStream(filePath);
+        if (jsonSchema != null) {
+            return jsonSchema;
+        }
+
+        if (schema.startsWith(CLASSPATH)) {
+            String filePath = schema.substring(CLASSPATH.length());
+            try (InputStream schemaStream = ValidationAspect.class.getResourceAsStream(filePath)) {
                 if (schemaStream == null) {
                     throw new IllegalArgumentException("'" + schema + "' is invalid, verify '" + filePath + "' is in your classpath");
                 }
+
                 jsonSchema = ValidationConfig.get().getFactory().getSchema(schemaStream);
-            } else {
-                jsonSchema = ValidationConfig.get().getFactory().getSchema(schema);
+            } catch (IOException e) {
+                throw new IllegalArgumentException("'" + schema + "' is invalid, verify '" + filePath + "' is in your classpath");
             }
+        } else {
+            jsonSchema = ValidationConfig.get().getFactory().getSchema(schema);
+        }
 
-            if (validateSchema) {
-                String version = ValidationConfig.get().getSchemaVersion().toString();
-                try {
-                    validate(jsonSchema.getSchemaNode(),
-                            getJsonSchema("classpath:/schemas/meta_schema_" + version));
-                } catch (ValidationException ve) {
-                    throw new IllegalArgumentException("The schema " + schema + " is not valid, it does not respect the specification " + version, ve);
-                }
+        if (validateSchema) {
+            String version = ValidationConfig.get().getSchemaVersion().toString();
+            try {
+                validate(jsonSchema.getSchemaNode(),
+                        getJsonSchema("classpath:/schemas/meta_schema_" + version));
+            } catch (ValidationException ve) {
+                throw new IllegalArgumentException("The schema " + schema + " is not valid, it does not respect the specification " + version, ve);
             }
-
-            schemas.put(schema, jsonSchema);
         }
 
+        schemas.put(schema, jsonSchema);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **231, 233, and 262**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **264**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.